timemachine/cpp/src/lambda_potential.cu

```diff
@@ -0,0 +1,154 @@
+    unsigned long long *out,
+    unsigned long long *buffer,
+    double lambda) {
+
+    int idx = blockIdx.x*blockDim.x + threadIdx.x;
+
+    if(idx >= count) {
+        return;
+    }
+
+    // convert fixed to double
+    double val = static_cast<double>(static_cast<long long>(buffer[idx]))/FIXED_EXPONENT;
+    val *= lambda;
+
+    unsigned long long fixed_val = static_cast<unsigned long long>((long long) (val*FIXED_EXPONENT));
+
+    atomicAdd(out + idx, fixed_val);
+
+}
+
+
+template<typename RealType>
+__global__ void k_reduce_add_buffer(
+    int count,
+    RealType *out,
+    RealType *buffer,
+    double lambda) {
+
+    int idx = blockIdx.x*blockDim.x + threadIdx.x;
+
+    if(idx >= count) {
+        return;
+    }
+
+    // does this work if one of them is an unsigned long long?
+    atomicAdd(out + idx, lambda*buffer[idx]);
+
+}
+
+__global__ void k_reduce_add_du_dl(
+    double *du_dl,
+    double *u_buf,
+    double *du_dl_buf,
+    double lambda) {
+
+    if(threadIdx.x == 0) {
+        atomicAdd(du_dl, u_buf[0] + lambda*du_dl_buf[0]);
+    }
+
+}
+
+
+LambdaPotential::LambdaPotential(
+    std::shared_ptr<Potential> u,
+    int N,
+    int P
+) : u_(u) {
+
+    gpuErrchk(cudaMalloc(&d_du_dx_buffer_, N*3*sizeof(*d_du_dx_buffer_)));
+    gpuErrchk(cudaMalloc(&d_du_dp_buffer_, P*sizeof(*d_du_dp_buffer_)));
+    gpuErrchk(cudaMalloc(&d_du_dl_buffer_, 1*sizeof(*d_du_dl_buffer_)));
+    gpuErrchk(cudaMalloc(&d_u_buffer_, 1*sizeof(*d_u_buffer_)));
+
+}
+
+LambdaPotential::~LambdaPotential() {
+    gpuErrchk(cudaFree(d_du_dx_buffer_));
+    gpuErrchk(cudaFree(d_du_dp_buffer_));
+    gpuErrchk(cudaFree(d_du_dl_buffer_));
+    gpuErrchk(cudaFree(d_u_buffer_));
+}
+
+void LambdaPotential::execute_device(
+    const int N,
+    const int P,
+    const double *d_x,
+    const double *d_p,
+    const double *d_box,
+    const double lambda,
+    unsigned long long *d_du_dx,
+    double *d_du_dp,
+    double *d_du_dl,
+    double *d_u,
+    cudaStream_t stream) {
+
+    // initialize buffer streams
+
+    if(d_du_dx) {
+       gpuErrchk(cudaMemsetAsync(d_du_dx_buffer_, 0, N*3*sizeof(*d_du_dx_buffer_), stream))
+    }
+
+    if(d_du_dp) {
+       gpuErrchk(cudaMemsetAsync(d_du_dp_buffer_, 0, P*sizeof(*d_du_dp_buffer_), stream))
+    }
+
+    if(d_du_dl) {
+        gpuErrchk(cudaMemsetAsync(d_du_dl_buffer_, 0, 1*sizeof(*d_du_dl_buffer_)));
+    }
+
+    // extra check is due to the chain rule
+    if(d_u || d_du_dl) {
+        gpuErrchk(cudaMemsetAsync(d_u_buffer_, 0, 1*sizeof(*d_u_buffer_)));
+    }
+
+    // let the alchemical potential a(l) be:
+    // a(l) = l*u(l)
+    // da/dl = u + l*du/dl
+    u_->execute_device(
+        N,
+        P,
+        d_x,
+        d_p,
+        d_box,
+        lambda,
+        d_du_dx ? d_du_dx_buffer_ : nullptr,
+        d_du_dp ? d_du_dp_buffer_ : nullptr,
+        d_du_dl ? d_du_dl_buffer_ : nullptr,
+        (d_du_dl || d_u) ? d_u_buffer_ : nullptr,
+        stream
+    );
+
+    int tpb = 32;
```

Comment:
This is equivalent to the warp size of all NVIDIA GPUs.

Reply:
I may be mixing up warps and blocks -- to clarify, is this number allowed to be a multiple of 32, or only exactly 32?

Reply:
The warp size is exactly 32 threads, that is the SIMD width of the GPU.

You are technically allowed to specify an arbitrary number of threads per block, however it's usually a multiple of 32. We use 32 here because nearly everything we do is via a rather advanced GPU technique called warp synchronous programming:

https://developer.nvidia.com/blog/using-cuda-warp-level